tpm2/structures.go

```diff
@@ -929,66 +1372,237 @@ type TPMSAuthCommand struct {
 // TPMSAuthResponse represents a TPMS_AUTH_RESPONSE.
 // See definition in Part 2: Structures, section 10.13.3.
 type TPMSAuthResponse struct {
+	marshalByReflection
 	Nonce         TPM2BNonce
 	Attributes    TPMASession
 	Authorization TPM2BData
 }
 
+type tpmuSymKeyBits struct {
+	selector TPMAlgID
+	contents Marshallable
+}
+
+type symKeyBitsContents interface {
+	TPMKeyBits | TPMAlgID
+}
+
+// create implements the UnmarshallableWithHint interface.
+func (u *tpmuSymKeyBits) create(hint int64) (reflect.Value, error) {
+	switch TPMAlgID(hint) {
+	case TPMAlgAES:
+		var contents boxed[TPMKeyBits]
+		u.contents = &contents
+		u.selector = TPMAlgID(hint)
+		return reflect.ValueOf(&contents), nil
+	case TPMAlgXOR:
+		var contents boxed[TPMAlgID]
+		u.contents = &contents
+		u.selector = TPMAlgID(hint)
+		return reflect.ValueOf(&contents), nil
+	}
+	return reflect.ValueOf(nil), fmt.Errorf("no union member for tag %v", hint)
+}
+
+// get implements the UnmarshallableWithHint interface.
+func (u *tpmuSymKeyBits) get(hint int64) (reflect.Value, error) {
+	if u.selector != 0 && hint != int64(u.selector) {
+		return reflect.ValueOf(nil), fmt.Errorf("incorrect union tag %v, is %v", hint, u.selector)
+	}
+	switch TPMAlgID(hint) {
+	case TPMAlgAES:
+		var contents boxed[TPMKeyBits]
+		if u.contents != nil {
+			contents = *u.contents.(*boxed[TPMKeyBits])
+		}
+		return reflect.ValueOf(&contents), nil
+	case TPMAlgXOR:
+		var contents boxed[TPMAlgID]
+		if u.contents != nil {
+			contents = *u.contents.(*boxed[TPMAlgID])
+		}
+		return reflect.ValueOf(&contents), nil
+	}
+	return reflect.ValueOf(nil), fmt.Errorf("no union member for tag %v", hint)
+}
+
 // TPMUSymKeyBits represents a TPMU_SYM_KEY_BITS.
 // See definition in Part 2: Structures, section 11.1.3.
-type TPMUSymKeyBits struct {
-	// TODO: The rest of the symmetric algorithms get their own entry
-	// in this union.
-	AES *TPMKeyBits  `gotpm:"selector=0x0006"` // TPM_ALG_AES
-	XOR *TPMIAlgHash `gotpm:"selector=0x000A"` // TPM_ALG_XOR
+func TPMUSymKeyBits[C symKeyBitsContents](selector TPMAlgID, contents C) tpmuSymKeyBits {
```

Comment:
note that the selector value is provided explicitly in this constructor, this is a little weird.

Ideally, a caller would be able to say "here's something that a TPMUFoo can hold, put it into a TPMUFoo for me" except that the TPMUFoo needs to somehow associate those contents with a selector value for future queries. It can't guess based on the concrete type of what was passed in (because you can't use generics to build function overloading, and because some TPMUs have multiple members with the same type).